storage: Fix circuit breaker flake in TestReplicateReAddAfterDown

a-robinson opened the pull request:
Fixes #10758. It's very possible that fixing the circuit breaker will cause other flakes in our code base to become more frequent, so feel free to point me at anything that seems to have gotten worse lately and I'll take a look. We may want to start injecting circuit breakers that won't trip into RaftTransport for the sake of avoiding flakes like this.

As for the mechanism of the fix, it might seem like it'd be better to add retries to [the call that's being blocked by the circuit breaker](https://github.com/cockroachdb/cockroach/blob/master/pkg/storage/client_test.go#L918), but we don't really know how many times to retry given that the circuit breaker's backoff is based on time rather than number of attempts.

@tamird

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10773)
<!-- Reviewable:end -->

a-robinson committed:
storage: Clean up use of magic constant in TestReplicateReAddAfterDown

petermattis commented on pkg/storage/client_raft_test.go:
```diff
@@ -2363,6 +2363,16 @@ func TestReplicateReAddAfterDown(t *testing.T) {
+	downedStoreIdx := 2
+	for _, transport := range mtc.transports {
+		transport.GetCircuitBreaker(roachpb.NodeID(downedStoreIdx + 1)).ShouldTrip = nil
+	}
```
I wonder if this should be done inside `multiTestContext` itself.

a-robinson replied:
Yeah, I was wondering that too. I played around with doing it for resetting circuit breakers before realizing that resetting them wasn't enough. I've moved this back into a helper.

petermattis reviewed (approved):
LGTM

petermattis commented on pkg/storage/client_test.go:
```diff
@@ -841,6 +841,18 @@ func (m *multiTestContext) restartStore(i int) {
+	for _, transport := range m.transports {
+		transport.GetCircuitBreaker(nodeID).ShouldTrip = nil
+	}
+}
```
This is cleaner, but I was actually wondering if we should always disable circuit breakers in `multiTestContext`-based tests. Is it useful to be testing them there? We already have this funky transport in `multiTestContext`.

a-robinson replied:
Ah, yeah. That's basically what I was wondering about in my PR description. It'd be a little clunky to do it like this, though, since we'd have to do it for the cross-product of nodes, and some nodes can be added after the multiTestContext is created.

We'd want to change the way the breakers are constructed. I was thinking that may involve a fair amount of work so I was just proposing it rather than doing it, but now I see that we already have a BreakerFactory in place :) Switched over to always disabling the breakers. I've been running `make stress` on the storage package for about five minutes now after making this change and still haven't run into any failures, so it looks like this may not have broken any tests' assumptions, which is a pleasant surprise.

a-robinson committed:
storage: Never trip circuit breakers in multiTestContext tests

This fixes/prevents timing-based flakes caused by the backoff logic in
circuit breakers.

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I see that `TestReportUnreachableHeartbeats` manually trips a circuit breaker. Does that still work? Seems like it should, just want to verify.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10773#-:-KWnnENLCgHXV15E83ti:b-2pj6uk)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Yeah, that still works just fine.

tamird commented:
Reviewed 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_test.go, line 849 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10773#-KWoJK14h-s10OdDSO2r:-KWoJK15htk3UkGOnJ45:b8qd689) ([raw file](https://github.com/cockroachdb/cockroach/blob/e01b5e1cbd376ed99b4a1dae9c3fe8d045e6b846/pkg/storage/client_test.go#L849)):*
> ```Go
> // ensure that errors from when the store was down don't cause other nodes to
> // avoid opening new connections to it.
> func (m *multiTestContext) disableCircuitBreakersToStore(i int) {
> ```

is this in the wrong commit? woof, and then it's gone in the next commit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10773)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_test.go, line 849 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10773#-KWoJK14h-s10OdDSO2r:-KWoLa-S4gtma5hUyNtc:b7ha7v4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e01b5e1cbd376ed99b4a1dae9c3fe8d045e6b846/pkg/storage/client_test.go#L849)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is this in the wrong commit? woof, and then it's gone in the next commit.
</details>
Whoops, my bad.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10773)*
<!-- Sent from Reviewable.io -->
